### PR TITLE
Dynamic theme improvements

### PR DESCRIPTION
**Pull request by artlowel**
## Description
This PR fixes an issue where if you have a different theme that's only active on certain routes, you can still see the old theme flash for a second first, before the correct theme is loaded. The cause was that the theme matching the route was determined in an ngrx effect, and effects are run in (something emulating) a different thread, so you could have some situations where the timings were off.

This PR addresses the issue by moving that logic to the theme service and triggering it from app component. That ensures the loading component is shown until the theme resolution is complete.

The PR also adds the ability for themes to extend each other, that way, if you want for example to have a shared root theme, and a slight variation for each community you don't need to copy the entire root theme to each community theme. The community themes can simply extend the root theme and only contain the components that differ.

## Instructions for Reviewers
You can test that the issue is fixed by configuring a visually distinct theme for a certain route in dspace, and going to that route by clicking a link in the UI (so not directly by pasting the URL) With this PR, you should always see a loading screen until the new theme is fully loaded.

You can test the extending of themes by creating a new theme (theme A) with one or two components, setting the `extends` property for that theme in the environment file to the name of another theme (theme B) and verifying that if theme A is active, the components in that theme are used, components that aren't in that fall back to the theme B versions.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

File: src/app/shared/metadata-representation/metadata-representation.decorator.spec.ts [hunk 1]
```diff
@@ -76,4 +96,44 @@ describe('MetadataRepresentation decorator function', () => {
       });
     });
   });
+
+  describe('With theme extensions', () => {
+    describe('If requested theme has no match', () => {
```
Comment (tdonohue) [inline src/app/shared/metadata-representation/metadata-representation.decorator.spec.ts:101]: Maybe I'm misunderstanding the theme extensions tests here...but, it seems like we are missing a test to prove extends is working?  I just see a test for "has no match" and "theme extension cycle".  Where's a test for a match?

UPDATE: It might be that this 'describe' is just badly named...it's just very hard to follow this test as-is without a bit more comments....I've had to try and review the beforeEach & the `it` clauses several times, and I'm still not sure it makes sense.

File: src/app/shared/object-collection/shared/listable-object/listable-object.decorator.spec.ts [hunk 2]
```diff
@@ -80,4 +100,44 @@ describe('ListableObject decorator function', () => {
...
+  describe('With theme extensions', () => {
+    describe('If requested theme has no match', () => {
+      beforeEach(() => {
+        environment.themes = [
+          { name: 'requested', extends: 'intermediate' },
+          { name: 'intermediate', extends: 'ancestor' },
+        ];
+      });
+
+      it('should return component from ancestor theme if it has a match', () => {
```
Comment (tdonohue) [inline src/app/shared/object-collection/shared/listable-object/listable-object.decorator.spec.ts:113]: I have the same questions about this `describe`->`it`. 

In other words, the sentence..."If requested theme has no match...it 'should return component from ancestor theme if it has a match'" doesn't make sense to me.  This appears to be trying to test for extending themes, but I'm not understanding the test description.

I again think some inline comments could be helpful here...it's hard to follow what `typeHier1` versus `typeHier2` means, and why they should return different components.

File: src/config/theme.model.ts [hunk 3]
```diff
@@ -6,6 +6,12 @@ import { getDSORoute } from '../app/app-routing-paths';
 // tslint:disable:max-classes-per-file
 export interface NamedThemeConfig extends Config {
   name: string;
+
+  /**
+   * Specify another theme to build upon: whenever a themed component is not found in the current theme,
+   * its ancestor theme(s) will be checked recursively before falling back to the default theme.
+   */
+  extends?: string;
```
Comment (tdonohue) [inline src/config/theme.model.ts:14]: I think this `extends` is a great concept & one that people will be excited about. But I think we need to **document it**.  Could we add an example or two to `environment.common.ts`?  

I'd also really appreciate it if you could add some basic docs to https://wiki.lyrasis.org/display/DSDOC7x/User+Interface+Customization about "Extending Other Themes"(with a note saying "This feature is coming in 7.1")

**Review by tdonohue — CHANGES_REQUESTED** [review]
@artlowel : Gave this a test today & it all seems to work.  I tried out the PR as-is, just to make sure I saw no obvious changes in behavior (none noticed).  I then tried out your sample branch to verify that the bug you noted is fixed...and it worked as well.

Overall, it looks good.  That said, I have a few minor requests... a couple specs have very confusing wording, either they need rewording or more inline comments.  And, I think this new `extends` option needs documentation (as I think this is a new feature we should advertise in 7.1).  Beyond that, it looks good!

File: src/app/shared/metadata-representation/metadata-representation.decorator.spec.ts (hunk 1 above)
Comment (ybnd) [reply src/app/shared/metadata-representation/metadata-representation.decorator.spec.ts:104]: The "there is a match" case is already covered by the existing tests; with these tests it should be enough to cover how we fall back to other themes if there is no match. I've added a little disclaimer comment.

Not sure how we can reword these`describe`s to make that more clear though...

File: src/app/shared/object-collection/shared/listable-object/listable-object.decorator.spec.ts (hunk 2 above)
Comment (ybnd) [reply src/app/shared/object-collection/shared/listable-object/listable-object.decorator.spec.ts]: Tried to reword the `it`s & explain a bit better

File: src/config/theme.model.ts (hunk 3 above)
Comment (ybnd) [reply src/config/theme.model.ts:14]: Added an example in comments.

File: src/app/shared/metadata-representation/metadata-representation.decorator.spec.ts (hunk 1 above)
Comment (tdonohue) [reply src/app/shared/metadata-representation/metadata-representation.decorator.spec.ts:104]: @ybnd : The renaming of variables and the `it` clauses helped make this a lot clearer. Thanks for that....my prior confusion was that the describe/it simply seemed contradictory, but with the updated language I see I had been misreading them. Thanks!

File: src/config/theme.model.ts (hunk 3 above)
Comment (tdonohue) [reply src/config/theme.model.ts:14]: @ybnd : Comments look good, thanks!

If you'd like a wiki account you can request one at wikihelp@lyrasis.org.  Once you have a wiki account created, I can give you access to edit the "DSDOC7x" area of the wiki.
